### DISCUSSION TITLE
Advise on proper comparison settings to use when data has spelling variations

### DISCUSSION THREAD

**Discussion by ali-beatty** [moj-analytical-services/splink#1799] [Q&A]
Been trying to figure out how to do proper configuration for this. Our data will often have names, no email addresses and job titles. Unfortunately, the job titles often change slightly when we get new data to merge
data example 

| first_name |last_name | job_title                                 |
|------------|-----------|-------------------------------------------|
| Frank      | Weiner    | Construction Utility Tech III             |
| Frank      | Weiner    | UTIL CONST TCH III                        |
| Bob        | Doe       | Customer Experience Librarian-Youth       |
| Robert     | Doe       | CUSTOMER EXPERIENCE LIBRARIAN - YOUTH     |

This configuration below works when there's an exact match, but that doesn't happen often.  

```python
settings = {
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        block_on("first_name"),
        block_on("last_name"),
    ],
    "comparisons": [
        ctl.name_comparison("first_name"),
        ctl.name_comparison("last_name"),
        cl.exact_match("job_title", term_frequency_adjustments=True),
    ],
```
Since I'm using postrgres I added
```python
    import splink.postgres.comparison_level_library as cll
```
and modified the settings to: 
```python
   "comparisons": [
        ctl.name_comparison("first_name"),
        ctl.name_comparison("last_name"),
        cll.levenshtein_level("job_title", 1)
```
but I get this error message: 
```<Comparison level 'Levenshtein <= 1' using SQL rule: levenshtein(... 
            is a comparison level and
            cannot be used as a standalone comparison.
```
I am sure I'm not configuring this properly, but unable to find examples similar to our use case, so also wondering if this is even possible. Attempting to match words that are the same, but using abbreviations and different spellings seems complicated to say the least. 

We currently use MS Excel's fuzzy mapping tool, and find it does a pretty good job on matching, but when there's a lot of variation in the job_title data it tends to not match.

Appreciate the support!

**Comment by ADBond** [comment]
To the specific error you are getting, this is because you are mixing **comparisons** and **comparison levels**. A comparison is made up of a collection of comparison levels, which define how you are dividing up the set of record pairs for that particular comparison. For instance, the default `name_comparison` settings will be made up of the following comparison levels:
* Exact match
* Damerau-Levenshtein Distance <= 1
* Jaro Winkler similarity >= 0.9
* Jaro Winkler similarity >= 0.8
* Anything else

[This topic guide on defining comparisons](https://moj-analytical-services.github.io/splink/topic_guides/comparisons/customising_comparisons.html) might help clarify things a bit further.

So your line `cll.levenshtein_level("job_title", 1)` doesn't make sense in this context - you need to define the _other_ levels of the `job_title` comparison which you will be assessing. The thing you would use here for a comparison with only Levenshtein levels would be [`splink.postgres.comparison_library.levenshtein_at_thresholds`](https://moj-analytical-services.github.io/splink/comparison_library.html#splink.comparison_library.LevenshteinAtThresholdsBase), which has levels e.g.:
* Exact match
* Levenshtein Distance <= 1
* Levenshtein Distance <= 2
* (you customise what distances you want in practice)
* Anything else

If you want to customise the levels that use further than this, _then_ you can use the `comparison_level_library` - see [the topic guide on different ways to define comparisons - building up from levels](https://moj-analytical-services.github.io/splink/topic_guides/comparisons/customising_comparisons.html#method-3-comparisonlevels).

**Comment by ADBond** [comment]
As to your actual data, it may prove useful to clean the `job_title` field before linkage - for example defining a new column `job_title_standardised` where e.g. text is all lowercase, remove punctuation, standardise spacing. This would mean that for instance your last two records would then be an exact match on this standardised field.
Depending on what the full data looks like it might even be worth further making a column which is an array of words (e.g. `['Construction', 'Utility', 'Tech', 'III']`, and then using [`array_intersect_at_sizes`](https://moj-analytical-services.github.io/splink/comparison_library.html#splink.comparison_library.ArrayIntersectAtSizesBase) - this would account for words appearing in different orders, but wouldn't take into account typos/alternate spellings without more customisation, so you might want something like (without thinking too carefully about this):
* exact match on `job_title_standardised`
* array intersect on `job_title_array` >= 3
* array intersect on `job_title_array` >= 1
* levenshtein on `job_title_standardised` <= 1
* levenshtein on `job_title_standardised` <= 3
* everything else

It will probably take a little bit of trial and error with your dataset to figure out what will work best for your use-case.
Let me know if anything is not clear or you have any further questions!
